lib/puppet/indirector/request.rb

```diff
@@ -197,19 +197,15 @@ def do_request(srv_service=:puppet, default_server=nil, default_port=nil, &block
     end
 
     # ... Fall back onto the default server.
-    begin
-      bound_server = Puppet.lookup(:server)
-    rescue
+    if bound_server = Puppet.lookup(:server) { nil }
```

Comment:
@joshcooper it looks like I didn't fully understand what you said to do here. This doesn't work, as it's getting `nil` as the server version, which is causing the run to fail

```
$ be ./bin/puppet agent -t --summarize
Warning: Unable to fetch my node definition, but the agent run will continue:
Warning: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Info: Retrieving pluginfacts
Error: /File[/Users/melissa/.puppetlabs/opt/puppet/cache/facts.d]: Failed to generate additional resources using 'eval_generate': Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Error: /File[/Users/melissa/.puppetlabs/opt/puppet/cache/facts.d]: Could not evaluate: Could not retrieve file metadata for puppet:///pluginfacts: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Info: Retrieving plugin
Error: /File[/Users/melissa/.puppetlabs/opt/puppet/cache/lib]: Failed to generate additional resources using 'eval_generate': Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Error: /File[/Users/melissa/.puppetlabs/opt/puppet/cache/lib]: Could not evaluate: Could not retrieve file metadata for puppet:///plugins: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Error: Could not retrieve catalog from remote server: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
Time:
         Last run: 1521833747
Version:
           Config:
           Puppet: 5.3.6
Error: Could not send report: Failed to open TCP connection to :0 (Can't assign requested address - connect(2) for nil port 0)
```

I'll dig into it more after lunch, but in the meantime if you have clarification, that'd be awesome

Reply:
Might try this instead (where the block is only yielded to if `server` hasn't been pushed on the context).


```ruby
bound_server = Puppet.lookup(:server) do
  if primary_server = Puppet.settings[:server_list][0]
    primary_server[0]
  else
    Puppet.settings[:server]
  end
end
```